### PR TITLE
Spark driver cluster mode for mesos

### PR DESCRIPTION
**Pull request by LeightonWong**
Related to #663 ，I break it into two pr. Without spark 2.0 upgrade, this pr only contain mesos cluster mode changes.
More detail usage, please refer to #608 .

### CODE REVIEW

File: job-server/src/main/scala/spark/jobserver/BinaryManager.scala [hunk 1]
```diff
@@ -40,7 +40,7 @@ class BinaryManager(jobDao: ActorRef) extends InstrumentedActor {
   import scala.concurrent.duration._
   import akka.pattern.{ask, pipe}
   import context.dispatcher
-  implicit val daoAskTimeout = Timeout(3 seconds)
+  implicit val daoAskTimeout = Timeout(60 seconds)
```
Comment (hntd187) [inline job-server/src/main/scala/spark/jobserver/BinaryManager.scala]: Why this change?

File: job-server/src/main/scala/spark/jobserver/JobCache.scala [hunk 2]
```diff
@@ -28,7 +31,7 @@ class JobCacheImpl(maxEntries: Int,
 
   private val cache = new LRUCache[(String, DateTime, String, BinaryType), BinaryJobInfo](maxEntries)
   private val logger = LoggerFactory.getLogger(getClass)
-  implicit val daoAskTimeout: Timeout = Timeout(3 seconds)
+  implicit val daoAskTimeout: Timeout = Timeout(60 seconds)
```
Comment (hntd187) [inline job-server/src/main/scala/spark/jobserver/JobCache.scala]: again

File: job-server/src/main/scala/spark/jobserver/BinaryManager.scala (hunk 1 above)
Comment (LeightonWong) [inline job-server/src/main/scala/spark/jobserver/BinaryManager.scala]: When we test mesos cluster mode, we often got Timeout error, because the jar we submit is a little big(about 30M) and our database is not fast enough,  and there are many synchronize call in spark-jobserver like `jobDao ? JobDAOActor.SaveBinary(appName, binaryType, uploadTime, binBytes)).` and `val resp = (jobDao ? JobDAOActor.GetApps(filterOpt)).mapTo[JobDAOActor.Apps]
      resp.map { msg => msg.apps } pipeTo requestor`, I simply enlarge the daoAskTimeout, but it is better to provide a config for this timeout.

File: job-server/src/main/scala/spark/jobserver/JobCache.scala (hunk 2 above)
Comment (LeightonWong) [inline job-server/src/main/scala/spark/jobserver/JobCache.scala]: As above, this `val jarBinaryReq = (dao ? JobDAOActor.GetBinaryContent(appName, BinaryType.Jar, uploadTime))
          .mapTo[JobDAOActor.BinaryContent]` in JobCache often cause timeout error, 3 seconds is not enough.

File: job-server/src/main/scala/spark/jobserver/BinaryManager.scala (hunk 1 above)
Comment (noorul) [inline job-server/src/main/scala/spark/jobserver/BinaryManager.scala]: I think we should make all timeouts configurable

**Comment by noorul** [comment]
Some unit tests are missing that is why the score got decreased.

**Comment by LeightonWong** [comment]
@noorul OK, I will try to add some unit tests to cover these diff.

**Comment by velvia** [comment]
@LeightonWong this LGTM as well.

@noorul @zeitos @hntd187 are we OK with the drop in code coverage?  Let's try to get this in before publishing 0.7

**Comment by noorul** [comment]
@LeightonWong Have you tested using dependent-jar-uris?

**Comment by LeightonWong** [comment]
@noorul Sorry, I did not test dependent-jar-uris, In cluster mode, JobManager started at some remote node other than the job-server running machine, I think if dependent-jar-uri use remote protocal like http://xxx.jar it will work, but if using file://xxx.jar, we have to put all the dependent jars in this machine local file system and it might be work also.

In our use case, we put all spark job third party dependent jars such like `alluxio-1.4.0-spark-client-jar-with-dependencies.jar` in spark `libs` directory with spark 1.6.x version and `jars` directory with spark 2.x version, so we are not using job-server's `dependent-jar-uris` config.

**Comment by velvia** [comment]
@LeightonWong sorry do you mind merging changes with master?  We can then submit this to master and release.   Thanks.

**Comment by LeightonWong** [comment]
@velvia Ok, no problem, I will merge master and resolve conflicts right now.

**Comment by velvia** [comment]
Thanks!

**Comment by velvia** [comment]
@LeightonWong thanks.  Can't see diff for some reason but Travis passes, so will merge now.

File: job-server/src/main/scala/spark/jobserver/AkkaClusterSupervisorActor.scala [hunk 3]
```diff
@@ -208,16 +208,16 @@ class AkkaClusterSupervisorActor(daoActor: ActorRef) extends InstrumentedActor {
...
     logger.info("Starting context with actor name {}", contextActorName)
 
-    val contextDir: java.io.File = try {
-        createContextDir(name, contextConfig, isAdHoc, contextActorName)
-      } catch {
-        case e: Exception =>
-          failureFunc(e)
-          return
-      }
+    val driverMode = Option(config.getString("spark.jobserver.driver-mode")).getOrElse("client")
```
Comment (noorul) [inline job-server/src/main/scala/spark/jobserver/AkkaClusterSupervisorActor.scala]: @LeightonWong This line breaks the existing functionality on master. I think there should be a default value for this in application.conf or you should handle the case that there might not be one.

Comment (LeightonWong) [inline job-server/src/main/scala/spark/jobserver/AkkaClusterSupervisorActor.scala]: Sorry, It's better to add a default value for driver-mode in application.conf, but will it break the functionality? If the config is not set, the default "client" mode will be used, and nothing change as before.

Comment (noorul) [inline job-server/src/main/scala/spark/jobserver/AkkaClusterSupervisorActor.scala]: @LeightonWong I agree! Let us add a default in application.conf
